### PR TITLE
Speed up cluster creation

### PR DESCRIPTION
**Pull request by lukeseawalker**
Separate the provisioning process of the ComputeNode
into two steps, boot and init. Then start provisioning (boot phase) both
MasterNode and ComputeNode at the same time, but wait to start the
init step for the ComputeNode until the Master has finished.

The technique used is described here
https://aws.amazon.com/it/blogs/devops/optimize-aws-cloudformation-templates/

No gain when initial_queue_size = 0

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -28,6 +28,10 @@ Please open a GitHub issue for any feedback or issues:
 https://github.com/awslabs/cfncluster.  There is also an active AWS
 HPC forum which may be helpful:https://forums.aws.amazon.com/forum.jspa?forumID=192.
 
+CfnCluster 1.6 IAM Change
+=========================
+Between CfnCluster 1.5.2 and 1.6.0 we made a change to the CfnClusterInstancePolicy that adds “cloudformation:DescribeStackResource” permissions. If you’re using a custom policy (e.g. you specify "ec2_iam_role" in your config) be sure it includes this new permission. See https://cfncluster.readthedocs.io/en/latest/iam.html
```
[the remark this answers is not in the document]
Comment (lukeseawalker) [reply README.rst]: ok

File: cloudformation/cfncluster.cfn.json [hunk 2]
```diff
@@ -2810,11 +2849,15 @@
...
-          "Count": "1",
-          "Timeout": "PT30M"
+      }
+    },
+    "MasterServerWaitCondition" : {
+      "Type" : "AWS::CloudFormation::WaitCondition",
+      "DependsOn" : ["MasterServer"],
+      "CreationPolicy" : {
+        "ResourceSignal" : {
+          "Timeout": "PT15M",
```
[the remark this answers is not in the document]
Comment (lukeseawalker) [reply cloudformation/cfncluster.cfn.json]: reverted to 30M

File: docs/source/iam.rst [hunk 3]
```diff
@@ -4,11 +4,12 @@ IAM in CfnCluster
 ========================
 
 .. warning::
+    Between CfnCluster 1.5.2 and 1.6.0 we added a change to the `CfnClusterInstancePolicy` that adds "cloudformation:DescribeStackResource" permissions. If you're using a custom policy (e.g. you specify "ec2_iam_role" in your config) be sure it includes this new permission.
```
[the remark this answers is not in the document]
Comment (lukeseawalker) [reply docs/source/iam.rst]: ok

**Comment by dougalb** [comment]
Overall comment. When there is a configuration error that prevents instances signaling to CFN, this approach will double the cost of failure vs the previous implementation. It was intentional and part of the early design to ensure that clusters did not incur more costs than needed during failure or early testing.

**Comment by lukeseawalker** [comment]
Hi @dougalb 
True. In case of error all the instances, master node and compute fleet, will be billed for the time they stay running, till the cloudformation timeout. We should also add that billing is now per-second and paying a bit more in case something is misconfigured sounds a good trade off to have a faster cluster deploy when everything is ok, which should be the usual situation.
As a mitigation, we could think to suggest to use the initial queue size equal to zero for early testing or to reduce the cloudformation timeout after which considering the instance creation failed (or even make this parameter configurable).

File: cloudformation/aws-parallelcluster.cfn.json [hunk 4]
```diff
@@ -3900,6 +3930,30 @@
                   "  vendor_cookbook\n",
                   "fi\n",
                   "cd /tmp\n",
+                  "\n",
+                  "masterServerStatus=$(aws cloudformation describe-stack-resource --stack-name ",
```
Comment (enrico-usai) [inline cloudformation/aws-parallelcluster.cfn.json]: What about using `Fn::Sub` to replace StackName and Region in the whole user-data script?

See **UserData Commands** section of the following documentation:
https://docs.aws.amazon.com/en_us/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-sub.html

(Then I know that is not part of this patch but we could also think to use `Fn::Join: ["\n", [...]]` to merge the multiple lines of the user-data script)

**Comment by lukeseawalker** [comment]
doc PR https://github.com/awsdocs/aws-parallelcluster-user-guide/pull/5

**Pull request merged**